// lingo3d/lib/api/settings.js
import { getFillWindow, setFillWindow } from "../states/useFillWindow";
import { getPixelRatio, setPixelRatio } from "../states/usePixelRatio";
import { getViewportSize, setViewportSize } from "../states/useViewportSize";
import { getDefaultFog, setDefaultFog } from "../states/useDefaultFog";
import { getDefaultLight, setDefaultLight } from "../states/useDefaultLight";
import { getGridHelper, setGridHelper } from "../states/useGridHelper";
import { getCameraHelper, setCameraHelper } from "../states/useCameraHelper";
import { getLightHelper, setLightHelper } from "../states/useLightHelper";
import { getPerformance, setPerformance } from "../states/usePerformance";
import { getOrbitControls, setOrbitControls } from "../states/useOrbitControls";
import { getResolution, setResolution } from "../states/useResolution";
export default {
    get fillWindow() {
        return getFillWindow();
    },
    set fillWindow(value) {
        setFillWindow(value);
    },
    get viewportSize() {
        return getViewportSize();
    },
    set viewportSize(val) {
        setViewportSize(val);
    },
    get resolution() {
        return getResolution();
    },
    set resolution(val) {
        setResolution(val);
    },
    get pixelRatio() {
        return getPixelRatio();
    },
    set pixelRatio(value) {
        setPixelRatio(value);
    },
    get performance() {
        return getPerformance();
    },
    set performance(value) {
        setPerformance(value);
    },
    get gridHelper() {
        return getGridHelper();
    },
    set gridHelper(value) {
        setGridHelper(value);
    },
    get cameraHelper() {
        return getCameraHelper();
    },
    set cameraHelper(value) {
        setCameraHelper(value);
    },
    get lightHelper() {
        return getLightHelper();
    },
    set lightHelper(value) {
        setLightHelper(value);
    },
    get defaultFog() {
        return getDefaultFog();
    },
    set defaultFog(value) {
        setDefaultFog(value);
    },
    get defaultLight() {
        return getDefaultLight();
    },
    set defaultLight(value) {
        setDefaultLight(value);
    },
    get defaultOrbitControls() {
        return getOrbitControls();
    },
    set defaultOrbitControls(value) {
        setOrbitControls(value);
    },
    wasmPath: "https://www.lingo3d.com/wasm/",
    autoMout: true
};
